credentials/generate-revocation-set.py

```diff
@@ -198,27 +205,98 @@ def fetch_crl_from_url(url: str, timeout: int) -> x509.CertificateRevocationList
         logging.error('Failed to fetch a valid CRL')
 
 
-class DCLDClient:
+class DCLDClientInterface:
     '''
-    A client for interacting with DCLD using either the REST API or command line interface (CLI).
+    An interface for interacting with DCLD.
 
     '''
 
-    def __init__(self, use_rest: bool, dcld_exe: str, production: bool, rest_node_url: str):
+    def get_revocation_points(self) -> list[dict]:
+        '''
+        Get revocation points from DCL
+
+        Returns
+        -------
+        list[dict]
+            List of revocation points
+        '''
+        raise NotImplementedError
+
+    def get_revocations_points_by_skid(self, issuer_subject_key_id) -> list[dict]:
+        '''
+        Get revocation points by subject key ID
+
+        Parameters
+        ----------
+        issuer_subject_key_id: str
+            Subject key ID
+
+        Returns
+        -------
+        list[dict]
+            List of revocation points
+        '''
+        raise NotImplementedError
+
+    def get_certificate(self, issuer_name_b64: str, akid_hex: str) -> tuple[bool, Optional[x509.Certificate]]:
+        '''
+        Get certificate from DCL
+        '''
+        raise NotImplementedError
+
+    def get_first_approved_certificate(self, response: dict, akid_hex: str) -> tuple[bool, Optional[x509.Certificate]]:
+        '''
+        Get first approved certificate from DCL resposne.
+        '''
+        if response["approvedCertificates"]["certs"][0]["pemCert"]:
+            issuer_certificate = x509.load_pem_x509_certificate(
+                bytes(response["approvedCertificates"]["certs"][0]["pemCert"], 'utf-8'))
+        else:
+            raise requests.exception.NotFound(f"No certificate found for {akid_hex}")
+        return response["approvedCertificates"]["certs"][0]["isRoot"], issuer_certificate
+
+    def get_paa_cert(self, initial_cert: x509.Certificate) -> Optional[x509.Certificate]:
+        """Get the PAA certificate for the CRL Signer Certificate."""
+        issuer_name_b64 = get_issuer_b64(initial_cert)
+        akid = get_akid(initial_cert)
+        if akid is None:
+            return
+        paa_certificate = None
+        while not paa_certificate:
+            try:
+                akid_hex = akid.hex().upper()
+                is_root, issuer_certificate = self.get_certificate(issuer_name_b64, akid_hex)
```

Comment:
Comment re: get_certificate(). I suppose the issuer name is Issuer DN. Can we name the variable as that? Also, I think we should b64 decode that and pass it around. One way to re-structure the functions would be to have get_issuer_b64 to get_issuer_dn where the function returns the issuer DN. WDYT?

Reply:
After some more cleanup i’ve decided to remove the get_issuer(subject)_b64 altogether. Instead the x509.name.Name object is used which has the rfc4514_string() method for logging purposes. I’ve added a method to the base class for converting to b64 that is used in get_approved_certificate.

Reply:
Sounds good.